File: code/C++/observer/statisticsdisplay.cpp
#include "statisticsdisplay.h"

StatisticsDisplay::StatisticsDisplay(WeatherData * const w) :maxTemp(-1), minTemp(-1), sumTemp(0), count(0)
{
	this->weatherData = w;
	this->weatherData->registerOserver(this);
	this->isObserver = true;
}
StatisticsDisplay::~StatisticsDisplay()
{
	std::cout << "free StatisticsDisplay" << std::endl;
}
void StatisticsDisplay::update()
{
	float temp = this->weatherData->getTemperature();
	if (maxTemp == -1 || maxTemp < temp)
		maxTemp = temp;
	if (minTemp == -1 || minTemp > temp)
		minTemp = temp;
	sumTemp += temp;
	count++;
	display();
}
void StatisticsDisplay::display()
{
	std::cout << "StatisticsDisplay maxTemp:" << maxTemp << "F, minTemp:" << minTemp << "F, avgTemp:" << (sumTemp / count) << "F" << std::endl;
}
void StatisticsDisplay::registerObserver()
{
	if (!this->isObserver) {
		this->weatherData->registerOserver(this);
		this->isObserver = true;
	}
}
void StatisticsDisplay::removeObserver()
{
	if (this->isObserver) {
		this->weatherData->removeOserver(this);
		this->isObserver = false;
	}
}

File: code/Java/observer/weather/StatisticsDisplay.java
package observer.weather;

import observer.display.DisplayElement;
import observer.observer.Observer;

public class StatisticsDisplay implements Observer, DisplayElement {
    private WeatherData weatherData;
    private float maxTemp;
    private float minTemp;
    private float sumTemp;
    private int count;
    private boolean isObserver;

    public StatisticsDisplay(WeatherData weatherData) {
        this.weatherData = weatherData;
        this.weatherData.registerObserver(this);
        this.isObserver = true;

        this.maxTemp = -1;
        this.minTemp = -1;
        this.sumTemp = 0;
        this.count = 0;
    }

    @Override
    public void display() {
        System.out.println("Statistics maxTemp:" + maxTemp + "F minTemp:" + minTemp + "F avgTemp:" + (sumTemp/count) + "F");
    }

    @Override
    public void update() {
        float temp = this.weatherData.getTemperature();
        if(this.maxTemp == -1 || this.maxTemp < temp)
            this.maxTemp = temp;
        if(this.minTemp == -1 || this.minTemp > temp)
            this.minTemp = temp;
        this.sumTemp += temp;
        this.count++;
        display();
    }

    @Override
    public void registerObserver() {
        if(!this.isObserver) {
            this.weatherData.registerObserver(this);
            this.isObserver = true;
        }
    }

    @Override
    public void removeObserver() {
        if(this.isObserver) {
            this.weatherData.removeObserver(this);
            this.isObserver = false;
        }
    }
}
